### PR TITLE
API refactoring

### PR DESCRIPTION
**Pull request by danni-m**

### CODE REVIEW

File: src/module.c [hunk 1]
```diff
@@ -25,6 +26,111 @@ int ReplySeriesRange(RedisModuleCtx *ctx, Series *series, api_timestamp_t start_
...
+    }
+    if (RMUtil_ArgExists("CHUNK_SIZE", argv, argc, 1) && RMUtil_ParseArgsAfter("CHUNK_SIZE", argv, argc, "l", maxSamplesPerChunk) != REDISMODULE_OK) {
+        return REDISMODULE_ERR;
+    }
+    return REDISMODULE_OK;
+}
+
+int _parseAggregationArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, api_timestamp_t *time_delta,
+                         int *agg_type) {
+    RedisModuleString * aggTypeStr = NULL;
```
Comment (yaeltzirulnikov) [inline src/module.c]: need to add argument number check here. Also, if there is something wrong with the parsing, for example time_delta is not numeric, we return here TSDB_NOTEXISTS and we don't check for this error value later, so the command works

File: src/module.c [hunk 2]
```diff
@@ -25,6 +26,111 @@ int ReplySeriesRange(RedisModuleCtx *ctx, Series *series, api_timestamp_t start_
...
+                    long long *retentionSecs, long long *maxSamplesPerChunk, size_t *labelsCount, Label **labels) {
+    *retentionSecs = TSGlobalConfig.retentionPolicy;
+    *maxSamplesPerChunk = TSGlobalConfig.maxSamplesPerChunk;
+    *labelsCount = 0;
+    *labels = parseLabelsFromArgs(ctx, argv, argc, labelsCount);;
+
+    if (RMUtil_ArgExists("RETENTION", argv, argc, 1) && RMUtil_ParseArgsAfter("RETENTION", argv, argc, "l", retentionSecs) != REDISMODULE_OK) {
+        return REDISMODULE_ERR;
+    }
+    if (RMUtil_ArgExists("CHUNK_SIZE", argv, argc, 1) && RMUtil_ParseArgsAfter("CHUNK_SIZE", argv, argc, "l", maxSamplesPerChunk) != REDISMODULE_OK) {
```
[the remark this answers is not in the document]
Comment (danni-m) [reply src/module.c]: @yaeltzirulnikov we decided to keep it in the code but not in the docs since it will be changed.

File: src/config.c [hunk 3]
```diff
@@ -27,13 +27,15 @@ int ReadConfig(RedisModuleString **argv, int argc) {
             return TSDB_ERROR;
         }
 
-        printf("loaded default compation policy: %s\n\r", policy_cstr);
+        printf("loaded default compaction policy: %s\n\r", policy_cstr);
         TSGlobalConfig.hasGlobalConfig = TRUE;
     }
 
     if (argc > 1 && RMUtil_ArgIndex("RETENTION_POLICY", argv, argc) >= 0) {
```
Comment (mnunberg) [inline src/config.c]: There's a much better argument parsing system that we have in RediSearch that replaces RMUtil_ParseArgs.. Might be useful to look at this later on. Also, suggest using the RETENTION_POLICY as a constant (e.g. macro)

File: src/module.c [hunk 4]
```diff
@@ -25,6 +26,111 @@ int ReplySeriesRange(RedisModuleCtx *ctx, Series *series, api_timestamp_t start_
...
+        return REDISMODULE_ERR;
+    }
+    return REDISMODULE_OK;
+}
+
+int _parseAggregationArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, api_timestamp_t *time_delta,
```
Comment (mnunberg) [inline src/module.c]: static

File: src/module.c [hunk 5]
```diff
@@ -25,6 +26,111 @@ int ReplySeriesRange(RedisModuleCtx *ctx, Series *series, api_timestamp_t start_
...
+        }
+
+        return TSDB_OK;
+    }
+
+    return TSDB_NOTEXISTS;
+
+}
+
+int parseAggregationArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, api_timestamp_t *time_delta,
```
Comment (mnunberg) [inline src/module.c]: static

File: src/module.c [hunk 6]
```diff
@@ -25,6 +26,111 @@ int ReplySeriesRange(RedisModuleCtx *ctx, Series *series, api_timestamp_t start_
...
+            RedisModule_ReplyWithError(ctx, "TSDB: Failed to retrieve aggregation class");
+            return TSDB_ERROR;
+        }
+        return TSDB_OK;
+    } else {
+        return result;
+    }
+}
+
+int parseRangeArguments(RedisModuleCtx *ctx, Series *series, int start_index, RedisModuleString **argv,
```
Comment (mnunberg) [inline src/module.c]: static

File: src/module.c [hunk 7]
```diff
@@ -151,37 +258,42 @@ int TSDB_queryindex(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
...
         return RedisModule_ReplyWithError(ctx, "TSDB: failed parsing labels");
     }
-    RMUtil_ParseArgs(argv, argc, query_count + 1, "ll", &start_ts, &end_ts);
 
     if (CountPredicateType(queries, (size_t) query_count, EQ) == 0) {
         return RedisModule_ReplyWithError(ctx, "TSDB: please provide at least one matcher");
     }
 
-    RedisModuleDict *result = QueryIndex(ctx, queries, query_count);
+    RedisModuleDict *result = QueryIndex(ctx, queries, (size_t) query_count);
```
Comment (mnunberg) [inline src/module.c]: why cast and not just declare query_count as size_t in the first place?

File: src/module.c [hunk 8]
```diff
@@ -151,37 +258,42 @@ int TSDB_queryindex(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
...
     api_timestamp_t start_ts, end_ts;
     api_timestamp_t time_delta = 0;
-    RedisModuleString * aggTypeStr = NULL;
-    AggregationClass *aggObject = NULL;
-    int agg_type = 0;
 
     if (argc < 4)
         return RedisModule_WrongArity(ctx);
-    int query_count = argc - 1 - 2; // 1 is for the command, 2 is for start_ts and end_ts
+    Series fake_series;
```
Comment (mnunberg) [inline src/module.c]: good practice to initialize to 0; Series = fake_series {0}; unless the structure is very large.

**Comment by danni-m** [comment]
@yaeltzirulnikov @mnunberg updated.

File: src/module.c [hunk 9]
```diff
@@ -25,6 +26,117 @@ int ReplySeriesRange(RedisModuleCtx *ctx, Series *series, api_timestamp_t start_
...
+    }
+    if (RMUtil_ArgExists("CHUNK_SIZE", argv, argc, 1) && RMUtil_ParseArgsAfter("CHUNK_SIZE", argv, argc, "l", maxSamplesPerChunk) != REDISMODULE_OK) {
+        return REDISMODULE_ERR;
+    }
+    return REDISMODULE_OK;
+}
+
+static int _parseAggregationArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, api_timestamp_t *time_delta,
+                         int *agg_type) {
+    RedisModuleString * aggTypeStr = NULL;
```
Comment (yaeltzirulnikov) [inline src/module.c]: RMUtil_ArgExists is deprecated, should use RMUtil_ArgIndex

File: src/module.c [hunk 10]
```diff
@@ -25,6 +26,117 @@ int ReplySeriesRange(RedisModuleCtx *ctx, Series *series, api_timestamp_t start_
...
+    int offset = RMUtil_ArgExists("AGGREGATION", argv, argc, 0);
```
Comment (yaeltzirulnikov) [inline src/module.c]: RMUtil_ArgExists is deprecated, should use RMUtil_ArgIndex

**Pull request merged**